beets/autotag/mb.py

```diff
@@ -466,6 +471,43 @@ def album_info(release):
         info.genre = '; '.join(g[0] for g in sorted(genres.items(),
                                                     key=lambda g: -g[1]))
 
+    # Additional online sources if configured and MusicBrainz provides them.
+    # FIXME This check is always True because of config_default.yaml.
+    if ('url_rels' in config['musicbrainz'].keys() and
+            release.get('url-relation-list')):
+        d_release_url, d_master_url = None, None
+        spotify_url, bandcamp_url = None, None
+        fetch_discogs, fetch_spotify, fetch_bandcamp = False, False, False
+        if config['musicbrainz']['url_rels']['discogs'].get():
+            fetch_discogs = True
+        if config['musicbrainz']['url_rels']['spotify'].get():
+            fetch_spotify = True
+
+        for url in release['url-relation-list']:
+            if fetch_discogs and url['type'] == 'discogs':
+                if 'release' in url['target']:
+                    log.debug('Found link to Discogs release via MusicBrainz')
+                    d_release_url = url['target']
+                if 'master' in url['target']:
+                    log.debug('Found link to Discogs Master release via '
+                              'MusicBrainz')
+                    d_master_url = url['target']
+            if fetch_spotify and url['type'] == 'spotify':
+                log.debug('Found link to Spotify album via MusicBrainz')
+                spotify_url = url['target']
+            if fetch_bandcamp and url['type'] == 'spotify':
+                log.debug('Found link to Bandcamp release via MusicBrainz')
+                spotify_url = url['target']
+        # We prefer a Discogs Release URL, but a Master URL is better than
+        # nothing. FIXME not sure if this is a good idea!
+        discogs_url = d_release_url if d_release_url else d_master_url
+        if discogs_url:
+            info.discogs_albumid = util.extract_discogs_id_regex(discogs_url)
+        if spotify_url:
+            info.spotify_albumid = "FIXME id extractor method?"
```

Comment:
Hi @arsaboo, before I dig into the Spotify plugin code, I'm sure you know by heart! :-) Is there an ID extractor method in the plugin similar to the extrac_discogs_id_regex() method? Thanks in advance!

Reply:
Hey @JOJ0....this is super cool. Being able to get all the details from MB would be awesome. 

The Spotify plugin uses the `get_id` method to obtain the id (and that method is defined in the plugins.py):
https://github.com/beetbox/beets/blob/b11df49705b877e3753c165a444f9529fe59f19a/beetsplug/spotify.py#L333

You should be able to use that method. BTW, I am assuming that you are checking if `spotify_albumid` is already populated before overwriting that field.

Reply:
I will try to use that function from outside the plugin. Not sure if it will work, had massive problems with that with the discogs_id_extractor method. How can I import the this `_get_id()` method from the plugin do you think? I'll try but with the discogs_id_extractor, "I got circular import errors" and in the end had to move that function out of the plugin to make use of it.

Reply:
And thanks for the reminder that checking if it is already populated! But how would I know if it is correct or the user wants to overwrite it when reimporting?? There is not really I way to know what the user wants other than believing what he configured: If he set the new config directive to yes, then it will be overwritten. At least that would have been my plan. What do you think?

```
musicbrainz:
    url_rels:
        spotify: yes
```
-> would mean: I want to overwrite it!

Also brainstorming here: "url_rels" in the config is not very self-explanatory (it's called like that in the muscibrainz json response). Do you have a better idea on how this config directive should be called? Thanks in advance!

Reply:
Spotify refers to them as `external_ids` - basically all external IDs (besides MB can be part of it, including ISRC and anything else that may be available). Maybe we can use the same terminology here and explain more in the docs. 

The default behavior should be no-overwrite. Currently, the `spotify_albumid` is only populated when the user has explicitly used the Spotify tagger and manually matched it. Thus, it does not make sense to overwrite this information unless the user explicitly requests it. I think we should use a flag `-f` (or a config entry) to force overwrite if the field is already populated. The Spotify plugin uses a similar flag. 

I did not write that part, but for the `_get_id()`, you will have to import `MetadataSourcePlugin`
https://github.com/beetbox/beets/blob/b11df49705b877e3753c165a444f9529fe59f19a/beetsplug/spotify.py#L34

Reply:
Thanks so much for your hints @arsaboo!

Using MetadataSourcePlugin._get_id method doesnt work like that:

```
  File "/home/jojo/git/beets/beets/autotag/mb.py", line 510, in album_info
    info.spotify_albumid = MetadataSourcePlugin._get_id('album', 123)
TypeError: _get_id() missing 1 required positional argument: 'id_'
```

When looking at the method, it's obvious. It is contained in an abstract class and the variable self._regex needs to be implemented in the source plugin that is based on `MetadataSourcePlugin: https://github.com/beetbox/beets/blob/b11df49705b877e3753c165a444f9529fe59f19a/beets/plugins.py#L706

So I can't just import from MetaDataSourcePlugin and expect things to work. This is what I'm missing from the spotify plugin: https://github.com/beetbox/beets/blob/b11df49705b877e3753c165a444f9529fe59f19a/beetsplug/spotify.py#L74

Reply:
So the next obvious step would be to try importing from the spotify beetsplug directly. Now I'm hitting exactly the same error as I had with the discogs plugin. circular import error. I think using stuff from a beets plugin outside of that plugin, is just not supported as it is implemented right now.

```
  File "/home/jojo/git/beets/beets/autotag/mb.py", line 29, in <module>
    from beetsplug import spotify
  File "/home/jojo/git/beets/beetsplug/spotify.py", line 32, in <module>
    from beets.autotag.hooks import AlbumInfo, TrackInfo
ImportError: cannot import name 'AlbumInfo' from partially initialized module 'beets.autotag.hooks' (most likely due to a circular import) (/home/jojo/git/beets/beets/autotag/hooks.py)
```

I don't really now how to proceed here. BTW, if you have a look at my previous commits, this was the reason I had to move the discogs_id_extractor method to the utils module....I'm still not sure if this is they way to go though....

Reply:
I think the difference between Spotify and the current PR is that Spotify is a plugin (inside `beetsplug`), whereas you are modifying the `mb.py` directly. I am just thinking out aloud, but wondering if it will be easier to just create another plugin (say `mbextend`) to add these features.

Reply:
Actually I thought that mb.py is a plugin by itself. I thought I recall a PR, where musicbrainz was refactored to be a source plugin on it's own like all the others? Did I dream that? Or was it never finished? Will investigte PR's....

Ok about your thought: If I create beetsplug/mbextend.py, don't you think that I will run in exactly the same circular dependency error again?

Maybe I could try if it theoretically would work to run a method of one plugin in another plugin like this: I use discogs as the source plugin, and then try importing with `from beetsplug import spotify`

Reply:
actually, you may not even need to import Spotify. The `get_id` function is part of `MetadataSourcePlugin` which can be directly imported.

Reply:
This is what I tried to tell you in this post: https://github.com/beetbox/beets/pull/4474#discussion_r980796263

It can’t work! MetadataSourcePlugin is missing self.id_regex. It’s not existing. This is to be implemented in the child class, in this case the SpotifyPlugin class. Have a look in MetadataSourcePlugin code!

Reply:
Exactly, and I am thinking that you can create a new `MBExtendPlugin` class in the new plugin that achieves the same. Maybe I am missing something.

Reply:
Yeah I think you miss the fact that it would be clumsy to having to reimplement the regex again in that new plugin. Double code fir the same thing! Dont you think it would be better if this regex or the id extractor function as a whole is moved to a place outside of the spotify plugin, so others can use it?

- Have a look at what I did wit the discogs regex function in my previous commits. To make it usable from outside the discogs plugin I had to move it away from the plugin. I chose the utils module. Now I can use it.

similar provlem with spotify regex. should I move it outside of the spotify plugin?? Only way I see currently if we dont want to copy code!

you see the problem now?